src/state.cc

```diff
@@ -159,6 +159,23 @@ bool State::AddDefault(StringPiece path, string* err) {
   return true;
 }
 
+bool State::PathExistsBetween(Edge* from, Edge* to) {
+  EdgeAdjacencyMap::iterator it =
+      edge_adjacency_memoization_.find(make_pair(from, to));
+  if (it != edge_adjacency_memoization_.end())
+    return it->second;
+  bool found = false;
+  for (size_t i = 0; i < to->inputs_.size(); ++i) {
+    Edge* e = to->inputs_[i]->in_edge();
+    if (e && ((e == from) || PathExistsBetween(from, e))) {
```

Comment:
```suggestion
    if (e && (e == from || PathExistsBetween(from, e))) {
```

Reply:
done (code moved to src/missing_deps.cc)